src/middlewared/middlewared/plugins/acme_protocol.py

```diff
@@ -0,0 +1,363 @@
+            id,
+            new
+        )
+
+        return await self._get_instance(id)
+
+    @accepts(
+        Int('id', required=True)
+    )
+    async def do_delete(self, id):
+
+        await self.middleware.call('certificate.delete_domains_authenticator', id)
+
+        return await self.middleware.call(
+            'datastore.delete',
+            self._config.datastore,
+            id
+        )
+
+    @accepts(
+        Dict(
+            'update_txt_record',
+            Int('authenticator', required=True),
+            Str('key', required=True),
+            Str('domain', required=True),
+            Str('challenge', required=True)
+        )
+    )
+    def update_txt_record(self, data):
+
+        authenticator = self.middleware.call_sync('dns.authenticator._get_instance', data['authenticator'])
+
+        return self.__getattribute__(
+            f'update_txt_record_{authenticator["authenticator"].lower()}'
+        )(
+            data['domain'],
+            messages.ChallengeBody.from_json(json.loads(data['challenge'])),
+            jose.JWKRSA.fields_from_json(json.loads(data['key'])),
+            **authenticator['attributes']
+        )
+
+    '''
+    Few rules for writing authenticator functions
+    1) The name must start with "update_txt_record_"
+    2) The authenticator name in function should be lowercase e.g "route53"
+    3) The first 3 arguments must be domain, challenge and key. Rest will be what the 
+       credentials are required for authenticating and nothing else 
+    4) In case update_txt_record is unsuccessful, ValidationErrors should be RAISED with appropriate
+       status/reason.
+    '''
+
+    def update_txt_record_route53(self, domain, challenge, key, access_key_id, secret_access_key):
+        session = boto3.Session(
+            aws_access_key_id=access_key_id,
+            aws_secret_access_key=secret_access_key
+        )
+        client = session.client('route53')
+        verrors = ValidationErrors()
+
+        # Finding zone id for the given domain
+        paginator = client.get_paginator('list_hosted_zones')
+        target_labels = domain.rstrip('.').split('.')
+        zones = []
+        try:
+            for page in paginator.paginate():
+                for zone in page['HostedZones']:
+                    if zone['Config']['PrivateZone']:
+                        continue
+
+                    candidate_labels = zone['Name'].rstrip('.').split('.')
+                    if candidate_labels == target_labels[-len(candidate_labels):]:
+                        zones.append((zone['Name'], zone['Id']))
+            if not zones:
+                verrors.add(
+                    'dns_authenticator_update_record.domain',
+                    f'Unable to find a Route53 hosted zone for {domain}'
+                )
+        except boto_exceptions.ClientError as e:
+            verrors.add(
+                'dns_authenticator_update_record.credentials',
```

Comment:
As we don't pass credentials here, we shouldn't raise validation errors for them. Use `CallError`.